### PR TITLE
Enrich Tensor Ops

### PR DESCRIPTION
**Pull request by workofart**
## Description
Enriched the Tensor class operations preparing for more complex operations in the future (e.g. convolution operations)

- Add `stack()`
- Add `cat()`
- Add `transpose()`, and `T` (alias)
- Add `pad()`
- Add `__iadd__()`
- Add `__getitem__()` and `__setitem()__`
- Add `shape` (alias)
- Updated `__mul__()` to handle scalar inputs correctly to get scalar gradients
- Updated `__add__()` to correctly accumulate gradients
- Updated `sum()` to handle gradient broadcasting correctly for N-dimensional matrices
- Updated `maximum()` to follow pytorch's implementation of splitting grade
- Move `_reduce_ops_backward` to the `_backward()` function of `mean()`
- Simplified __matmul__() to handle N-dimensional matrix multiplication by leveraging numpy's `matmul` function with `swapaxes`
- changes to all tensor ops
  - updated all the `prev` to use set instead of tuple
  - check for requires_grad before accumulating gradients for `self` or `other`
- Updated `Tensor.backward()` to lazy initialize the gradients inside here, instead of in the Tensor `__init__` constructor

## Tests
Added associated unit tests for each change.

Verified all the training examples are working as expected.

### CODE REVIEW

File: test/autograd/test_train.py [hunk 1]
```diff
@@ -45,6 +45,6 @@ def test_binary_classification(self):
         y_pred = model(X).data
 
         # compare y_pred and y on the classification accuracy
-        accuracy = sum((y_pred > 0.5).astype(int) == y) / X.shape[0]
```
Comment (workofart) [inline test/autograd/test_train.py:48]: y_pred is now (N, 1) due to all gradients are now initialized as vectors/matrices instead of scalars, so we need to squeeze it to compute accuracy
